Properly report retry counts when refreshing the lead broker

faec opened the pull request:
This is a fix for https://github.com/elastic/beats/issues/19015, in which Sarama fails to apply exponential backoff when `Producer.Retry.BackoffFunc` is set to an exponential backoff function.

Sarama tracks producer retries per-message, and updates them when receiving a response from the broker. This means that when there is no broker, the retry count never gets incremented. Thus, retries of send attempts to a valid broker will properly apply the exponential backoff, but retries while connecting to the broker itself will always retry with the initial backoff time.

I have modified the wait so that it tracks retries on the broker itself separately from retries on its pending messages. Local tests as described in https://github.com/elastic/beats/issues/19015 confirm that exponential backoff is now correctly applied when the broker is down.

ycombinator commented:
Thanks for the explanation; I understand the change. Is this a chance we'd have to permanently keep in our fork, meaning it's just how _we_ want sarama to work OR is this potentially a bug in sarama itself whose fix we should try push upstream as well?

faec commented:
My expectation is that Sarama will consider this a bug and want to fix it, though possibly not the exact same way I did. (There is also possible data loss in the existing code when a broker is down, but I don't know Sarama well enough to evaluate that part, it might just be handled elsewhere.) I'll be filing a bug with them shortly. Our hope is still to drop this fork entirely once our blockers are addressed in the current release, but this does add one more blocker to the list.

I've been testing this PR more and I found a reproducible crash with the fix applied when a lead broker goes down and then comes back up, so evidently something isn't quite right yet -- I will follow up here once I track it down.

faec commented:
Fixed (knock on wood); the problem was that `pp.parent.returnError` returns an immediate fatal error for that message, and thus should only be called once per message (rather than in the error loop as in my first attempt).

I adjusted this PR to combine the old and new approaches: I still track leader-update retries separately from message retries for determining the wait time, but I also report a missing leader as an immediate error for that message and continuer to the next loop iteration, rather than blocking until I find one. This ensures progress while still reporting relevant errors and applying the correct backoff.

urso commented:
Given the description for the fix I wonder if it would make sense from sarama as a library point of view to configure the retry backoff for the leader separately from the backoff for messages. Having different configs would still allow us to configure our backoff handler to get this behavior.

Messages can be 'rescheduled' in sarama, like send to another partition after an error. But is there a good reason to actually have backoff at the message level? Normally I'd expect errors to be at the broker level (IO error). On the other hand, errors at the broker level + backoff can bring all publishing to halt (depends on config, though).
